### PR TITLE
Push built package recipes to OBS on release and generate changelogs

### PR DESCRIPTION
**Pull request by willprice**
This PR aims to close off the final part of our packaging pipeline: to automate the creation of packaging repositories for debian and RHEL based distributions (Including Fedora and CentOS). A few things needed to happen to enable this. 
1. I needed some way of generating a changelog for inclusion with the debian package. Prior to this PR, this would have to be manually updated in `packaging/deb/changelog`. It seemed stupid to duplicate the changelog in markdown for the GitHub release notes as well as in a different format for the debian package. Consequently, I defined an intermediate format (exemplified in `CHANGELOG.yml`) which is the canonical changelog source. This is to be updated prior to each release with notes on what has changed. I wrote the tool to generate the markdown and debian changelogs in python, making use of jinja2.
2. Since the changelog file included in the debian package are now dynamically generated, I had to add a new build step to the Makefile to generate the debian `changelog`  file and make the deb packages dependent upon this.
3. I've added a new build step to generate a debian *source* package, this is built from the debian packaging recipe and the source tarball. One of the outputs of this process is a `.dsc` file which is consumed by OBS to build the deb packages.

You might wonder why I don't generate changelogs for the `spec` RPM file. It transpires that in the RPM world, this changelog is for *packaging*, and not for the software. Therefore we should manually manage the changelog in the spec file, rather than generating it from `CHANGELOG.yml`.

You will also probably note that this is pushing to the `home:willprice:poplog` project on OBS. I've still not had any luck in contacting anyone to make me a project that is not under my user namespace.
I plan on updating this PR with either the new project, if we get one, else I'll create a new user on OBS using the getpoplog gmail address and we can just use that instead.

Built packages can be found at https://build.opensuse.org/package/show/home:getpoplog/poplog

This PR is dependent upon https://github.com/GetPoplog/Seed/pull/69 being merged first.

~**DO NOT MERGE, EVEN AFTER REVIEW**: we need a consistent packaging source on OBS, that either needs to be a shared project across multiple OBS users, or a project that is hosted under a user that is not tied to any single person's identity (i.e. we'll create a getpopolog user on OBS)~

~**IMPORTANT WARNING: Prior to merging this in, we need to add the filters to the `push_to_open_build_service1` so that it only runs on tagged commits.**~

### CODE REVIEW

**Comment by willprice** [comment]
@sfkleach, the opensuse builds fail due to some packaging checks that highlight a bunch of missing symlink targets that originate from the poplog packages tar: https://build.opensuse.org/package/live_build_log/home:willprice/poplog/SLE_12_Backports/x86_64
```
[  114s] + /usr/lib/rpm/brp-suse
[  114s] calling /usr/lib/rpm/brp-suse.d/brp-05-permissions
[  114s] setting / to root:root 0755. (wrong owner/group abuild:abuild)
[  114s] setting /opt/ to root:root 0755. (wrong owner/group abuild:abuild)
[  114s] setting /usr/ to root:root 0755. (wrong owner/group abuild:abuild)
[  114s] calling /usr/lib/rpm/brp-suse.d/brp-15-strip-debug
[  115s] calling /usr/lib/rpm/brp-suse.d/brp-25-symlink
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/setup/man -> /opt/poplog/V16/man
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/setup/bin -> /opt/poplog/V16/man/bin
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/newkit/prb/prb -> /opt/poplog/V16/pop/packages/newkit/prb/newkit/prb
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/LINK_MOTIF_POPLOG -> /opt/poplog/V16/INSTALL/LINK_MOTIF_POPLOG
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/CHECK_LINUX_FACILITIES -> /opt/poplog/V16/INSTALL/CHECK_LINUX_FACILITIES
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/LINK_LIKE_CORE -> /opt/poplog/V16/INSTALL/LINK_LIKE_CORE
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/startup/poplog.sh -> /opt/poplog/V16/bin/poplog.sh
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/startup/poplog -> /opt/poplog/V16/bin/poplog
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/startup/demos -> /opt/poplog/V16/bin/demos
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/LINK_USING_NEWPOP -> /opt/poplog/V16/INSTALL/LINK_USING_NEWPOP
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/LINK_LIKE_BHAM -> /opt/poplog/V16/INSTALL/LINK_LIKE_BHAM
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/LINK_NOBHAM_POPLOG -> /opt/poplog/V16/INSTALL/LINK_NOBHAM_POPLOG
[  115s] Add the package providing the target to BuildRequires and Requires
[  115s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  115s]   /opt/poplog/V16/pop/packages/com/LINK_NOMOTIF_POPLOG -> /opt/poplog/V16/INSTALL/LINK_NOMOTIF_POPLOG
[  115s] Add the package providing the target to BuildRequires and Requires
[  116s] ERROR: link target doesn't exist (neither in build root nor in installed system):
[  116s]   /opt/poplog/V16/pop/help/local -> /opt/poplog/local/help
[  116s] Add the package providing the target to BuildRequires and Requires
[  116s] INFO: relinking /opt/poplog/V16/pop/help/pack -> ../packages/teaching/help (was ../packages/teaching/help/)
[  116s] error: Bad exit status from /var/tmp/rpm-tmp.5bgy5C (%install)
```

**Comment by willprice** [comment]
I've written up some notes and linked to a bunch of resources used whilst doing this packaging work: https://github.com/GetPoplog/Seed/wiki/Packaging

**Comment by willprice** [comment]
I've switched the deploy user on OBS now to `getpoplog`: https://build.opensuse.org/package/show/home:getpoplog/poplog

**Comment by willprice** [comment]
@sfkleach I've added in the gating mechanism, so pushing to OBS will only run on release. This commit is ready to be merged post review.

**Comment by willprice** [comment]
Pretty instructions for users to add repos are here:
https://software.opensuse.org/download.html?project=home:getpoplog&package=poplog

File: Makefile [hunk 1]
```diff
@@ -527,44 +528,74 @@ relink-and-build:
...
-# Assumes the following dependencies are installed
-# - debmake
-# - debhelper
-_build/artifacts/poplog_$(GETPOPLOG_VERSION)-1_amd64.deb: $(SRC_TARBALL)
-	mkdir -p "$(@D)"
+.PHONY: debsrc
+debsrc: _build/artifacts/poplog_$(GETPOPLOG_VERSION)-1.dsc _build/artifacts/poplog_$(GETPOPLOG_VERSION)-1.tar.gz
+
+_build/packaging/deb/poplog-$(GETPOPLOG_VERSION): $(SRC_TARBALL) _build/changelogs/CHANGELOG.debian
+	mkdir -p "$@"
```
Comment (sfkleach) [inline Makefile:555]: `mkdir -p "$@/debian"` maybe?

**Review by sfkleach — APPROVED** [review]

**Pull request merged**